pandas/tests/indexes/multi/test_unique_and_duplicates.py

```diff
@@ -194,46 +207,155 @@ def f(a):
     n, k = 200, 5000
     levels = [np.arange(n), tm.makeStringIndex(n), 1000 + np.arange(n)]
     labels = [np.random.choice(n, k * n) for lev in levels]
-    mi = MultiIndex(levels=levels, labels=labels)
+    mi = pd.MultiIndex(levels=levels, labels=labels)
+
+    result = mi.duplicated(keep=keep)
+    expected = pd._libs.hashtable.duplicated_object(mi.values, keep=keep)
+    tm.assert_numpy_array_equal(result, expected)
 
-    for keep in ['first', 'last', False]:
-        left = mi.duplicated(keep=keep)
-        right = pd._libs.hashtable.duplicated_object(mi.values, keep=keep)
-        tm.assert_numpy_array_equal(left, right)
 
+def test_get_duplicates():
     # GH5873
     for a in [101, 102]:
-        mi = MultiIndex.from_arrays([[101, a], [3.5, np.nan]])
+        mi = pd.MultiIndex.from_arrays([[101, a], [3.5, np.nan]])
         assert not mi.has_duplicates
 
         with warnings.catch_warnings(record=True):
             # Deprecated - see GH20239
-            assert mi.get_duplicates().equals(MultiIndex.from_arrays(
+            assert mi.get_duplicates().equals(pd.MultiIndex.from_arrays(
                 [[], []]))
 
-        tm.assert_numpy_array_equal(mi.duplicated(), np.zeros(
-            2, dtype='bool'))
+        tm.assert_numpy_array_equal(mi.duplicated(),
+                                    np.zeros(2, dtype='bool'))
 
     for n in range(1, 6):  # 1st level shape
         for m in range(1, 5):  # 2nd level shape
             # all possible unique combinations, including nan
             lab = product(range(-1, n), range(-1, m))
-            mi = MultiIndex(levels=[list('abcde')[:n], list('WXYZ')[:m]],
-                            labels=np.random.permutation(list(lab)).T)
+            mi = pd.MultiIndex(levels=[list('abcde')[:n], list('WXYZ')[:m]],
+                               labels=np.random.permutation(list(lab)).T)
             assert len(mi) == (n + 1) * (m + 1)
             assert not mi.has_duplicates
 
             with warnings.catch_warnings(record=True):
                 # Deprecated - see GH20239
-                assert mi.get_duplicates().equals(MultiIndex.from_arrays(
+                assert mi.get_duplicates().equals(pd.MultiIndex.from_arrays(
                     [[], []]))
 
-            tm.assert_numpy_array_equal(mi.duplicated(), np.zeros(
-                len(mi), dtype='bool'))
+            tm.assert_numpy_array_equal(mi.duplicated(),
+                                        np.zeros(len(mi), dtype='bool'))
+
+
+@pytest.mark.parametrize('keep', ['first', 'last'])
+def test_duplicated_inverse(keep):
+    # check that return_inverse kwarg does not affect outcome;
+    # index of inverse must be correctly transformed as well
+    major_axis = pd.Index(['foo', 'bar'])
+    minor_axis = pd.Index(['one', 'two'])
+
+    major_labels = np.array([0, 0, 1, 0, 1, 1])
+    minor_labels = np.array([0, 1, 0, 1, 0, 1])
+    index_names = ['first', 'second']
+    mi = pd.MultiIndex(levels=[major_axis, minor_axis],
```

Comment:
Since this is shared across quite a few tests should make a module fixture

Reply:
I agree. This would have been something for #21514, but I'll pick it up.

Reply:
Done (git mistakenly does not show this as outdated).